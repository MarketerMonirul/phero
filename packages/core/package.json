{
  "name": "@samen/core",
  "version": "0.2.3",
  "main": "dist/index.js",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "tsc",
    "test": "jest src",
    "prepublish": "npm run build"
  },
  "devDependencies": {
    "@tsconfig/recommended": "^1.0.1",
    "@types/jest": "^26.0.18",
    "@types/node": "^14.14.12",
    "@types/prettier": "^2.1.5",
    "jest": "^26.6.3",
    "ts-jest": "^26.4.4"
  },
  "dependencies": {
    "ora": "^5.1.0",
    "prettier": "^2.2.0",
    "ts-morph": "^9.1.0",
    "typescript": "^4.1.2"
  }
}
