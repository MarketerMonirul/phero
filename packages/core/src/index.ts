export * from "./domain"
export * from "./errors"
export * as paths from "./paths"
export * from "./utils"
export * from "./transpiler"
