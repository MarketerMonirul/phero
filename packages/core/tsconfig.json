{
  "extends": "@tsconfig/recommended/tsconfig.json",
  "compilerOptions": {
    "outDir": "./dist",
    "declaration": true,
    "target": "ES2019",
    "lib": ["ES2020"]
  },
  "include": ["src"],
  "exclude": ["src/**/*.test.ts", "src/**/*.spec.ts"]
}
