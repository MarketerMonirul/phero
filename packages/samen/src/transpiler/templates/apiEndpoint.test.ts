import { RPCFunction, JSType, render } from "@samen/core"
import apiEndpoint, { Props } from "./apiEndpoint"

const props = (partialRpcFunction: Partial<RPCFunction>): Props => {
  const rpcFunction: RPCFunction = {
    name: "testFunction",
    parameters: [],
    returnType: { type: JSType.untyped },
    modelIds: [],
    namespace: [],
    ...partialRpcFunction,
  }
  return {
    rpcFunction,
    manifest: {
      rpcFunctions: [rpcFunction],
      models: {
        Something: {
          name: "Something",
          namespace: [],
          id: "Something",
          ts: "interface Something { x: number }",
        },
      },
      refs: {
        Something: {
          id: "Something",
          modelId: "Something",
          value: { type: JSType.ref, id: "Something" },
        },
      },
    },
    relativeSamenFilePath: "../samen",
    config: {},
  }
}

test("literal parameter and return type", () => {
  expect(
    render(
      apiEndpoint(
        props({
          parameters: [{ name: "a", index: 0, value: { type: JSType.number } }],
          returnType: { type: JSType.string },
        }),
      ),
    ),
  ).toMatchSnapshot()
})

test("ref parameter and return type", () => {
  expect(
    render(
      apiEndpoint(
        props({
          parameters: [
            {
              name: "a",
              index: 0,
              value: { type: JSType.ref, id: "Something" },
            },
          ],
          returnType: { type: JSType.ref, id: "Something" },
        }),
      ),
    ),
  ).toMatchSnapshot()
})
