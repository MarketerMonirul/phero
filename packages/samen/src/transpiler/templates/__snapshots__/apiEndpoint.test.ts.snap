// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`literal parameter 1`] = `
"import { testFunction } from \\"../samen\\";

export interface ValidationError {
  scope: string;
  jsValue: unknown;
  message: string;
}

class InvalidInputError extends Error {
  errorCode = \\"INVALID_INPUT_ERROR\\";
  constructor(private readonly errors: ValidationError[]) {
    super(\\"Invalid input for RPC\\");
  }
}

export function validate(a: unknown): ValidationError[] {
  function validateTypeOf(
    typeString: \\"string\\" | \\"number\\" | \\"boolean\\",
    scope: string,
    jsValue: unknown
  ): ValidationError[] {
    return typeof jsValue === typeString
      ? []
      : [{ scope, jsValue, message: \`value is not a \${typeString}\` }];
  }

  function validateValueOf(
    possibleValues: any[],
    scope: string,
    jsValue: any
  ): ValidationError[] {
    return possibleValues.includes(jsValue)
      ? []
      : [
          {
            scope,
            jsValue,
            message: \`value must be one of: \${possibleValues
              .map((v) => \`\\"\${v}\\"\`)
              .join(\\", \\")}\`,
          },
        ];
  }

  function validateDate(scope: string, jsValue: unknown): ValidationError[] {
    const r = /^d{4}-dd-ddTdd:dd:dd(.d+)?(([+-]dd:dd)|Z)?$/i;
    const isValidDate =
      jsValue instanceof Date ||
      (typeof jsValue === \\"string\\" &&
        r.test(jsValue) &&
        !isNaN(Date.parse(jsValue)));
    return isValidDate
      ? []
      : [{ scope, jsValue, message: \\"value must be a date\\" }];
  }

  function validateNull(scope: string, jsValue: unknown): ValidationError[] {
    return jsValue === null
      ? []
      : [{ scope, jsValue, message: \\"value must be null\\" }];
  }

  function validateUndefined(
    scope: string,
    jsValue: unknown
  ): ValidationError[] {
    return jsValue === undefined
      ? []
      : [{ scope, jsValue, message: \\"value must be undefined\\" }];
  }

  function validateObject(scope: string, jsValue: unknown): ValidationError[] {
    return jsValue !== null && typeof jsValue === \\"object\\"
      ? []
      : [{ scope, jsValue, message: \\"value must be an object\\" }];
  }

  function validateArray(scope: string, jsValue: unknown): ValidationError[] {
    return Array.isArray(jsValue)
      ? []
      : [{ scope, jsValue, message: \\"value must be an array\\" }];
  }

  function validateTuple(
    length: number,
    scope: string,
    jsValue: unknown
  ): ValidationError[] {
    return Array.isArray(jsValue) && jsValue.length === length
      ? []
      : [
          {
            scope,
            jsValue,
            message: \`value must be a tuple of length \${length}\`,
          },
        ];
  }

  function validateOneOfTypes(
    numberOfTypes: number,
    scope: string,
    jsValue: unknown,
    isValid: boolean
  ): ValidationError[] {
    return isValid
      ? []
      : [
          {
            scope,
            jsValue,
            message: \`value is neither of the \${numberOfTypes} union types\`,
          },
        ];
  }

  function flatten<T>(arr: T[][]): T[] {
    return arr.reduce((r, a) => [...r, ...a], []);
  }

  const refs: { [refId: string]: (jsValue: unknown) => ValidationError[] } = {};
  refs[\`Something\`] = (jsValue: unknown): ValidationError[] =>
    refs[\`Something\`](jsValue);

  return [...validateTypeOf(\\"number\\", \\"a\\", a)];
}

export async function handler(event: any) {
  const body = JSON.parse(
    event.body === null || event.body === undefined ? \\"{}\\" : event.body
  );

  const inputValidationResult = validate(body.a);

  if (inputValidationResult.length) {
    return {
      statusCode: 400,
      body: JSON.stringify(
        new InvalidInputError(inputValidationResult),
        null,
        4
      ),
      headers: {
        \\"Content-Type\\": \\"application/json\\",
      },
    };
  }

  try {
    const result = await testFunction(body.a);

    return {
      statusCode: 200,
      body: null,
      headers: {
        \\"Content-Type\\": \\"application/json\\",
      },
    };
  } catch (e) {
    return {
      statusCode: 500,
      body: JSON.stringify(e, null, 4),
      headers: {
        \\"Content-Type\\": \\"application/json\\",
      },
    };
  }
}

export async function rpc_testFunction(a: number): Promise<void> {
  const inputValidationResult = validate(a);

  if (inputValidationResult.length) {
    throw new InvalidInputError(inputValidationResult);
  }

  const result = await testFunction(a);
  return result;
}
"
`;

exports[`literal parameter and return type 1`] = `
"import { testFunction } from \\"../samen\\";

export interface ValidationError {
  scope: string;
  jsValue: unknown;
  message: string;
}

class InvalidInputError extends Error {
  errorCode = \\"INVALID_INPUT_ERROR\\";
  constructor(private readonly errors: ValidationError[]) {
    super(\\"Invalid input for RPC\\");
  }
}

export function validate(a: unknown): ValidationError[] {
  function validateTypeOf(
    typeString: \\"string\\" | \\"number\\" | \\"boolean\\",
    scope: string,
    jsValue: unknown
  ): ValidationError[] {
    return typeof jsValue === typeString
      ? []
      : [{ scope, jsValue, message: \`value is not a \${typeString}\` }];
  }

  function validateValueOf(
    possibleValues: any[],
    scope: string,
    jsValue: any
  ): ValidationError[] {
    return possibleValues.includes(jsValue)
      ? []
      : [
          {
            scope,
            jsValue,
            message: \`value must be one of: \${possibleValues
              .map((v) => \`\\"\${v}\\"\`)
              .join(\\", \\")}\`,
          },
        ];
  }

  function validateDate(scope: string, jsValue: unknown): ValidationError[] {
    const r = /^d{4}-dd-ddTdd:dd:dd(.d+)?(([+-]dd:dd)|Z)?$/i;
    const isValidDate =
      jsValue instanceof Date ||
      (typeof jsValue === \\"string\\" &&
        r.test(jsValue) &&
        !isNaN(Date.parse(jsValue)));
    return isValidDate
      ? []
      : [{ scope, jsValue, message: \\"value must be a date\\" }];
  }

  function validateNull(scope: string, jsValue: unknown): ValidationError[] {
    return jsValue === null
      ? []
      : [{ scope, jsValue, message: \\"value must be null\\" }];
  }

  function validateUndefined(
    scope: string,
    jsValue: unknown
  ): ValidationError[] {
    return jsValue === undefined
      ? []
      : [{ scope, jsValue, message: \\"value must be undefined\\" }];
  }

  function validateObject(scope: string, jsValue: unknown): ValidationError[] {
    return jsValue !== null && typeof jsValue === \\"object\\"
      ? []
      : [{ scope, jsValue, message: \\"value must be an object\\" }];
  }

  function validateArray(scope: string, jsValue: unknown): ValidationError[] {
    return Array.isArray(jsValue)
      ? []
      : [{ scope, jsValue, message: \\"value must be an array\\" }];
  }

  function validateTuple(
    length: number,
    scope: string,
    jsValue: unknown
  ): ValidationError[] {
    return Array.isArray(jsValue) && jsValue.length === length
      ? []
      : [
          {
            scope,
            jsValue,
            message: \`value must be a tuple of length \${length}\`,
          },
        ];
  }

  function validateOneOfTypes(
    numberOfTypes: number,
    scope: string,
    jsValue: unknown,
    isValid: boolean
  ): ValidationError[] {
    return isValid
      ? []
      : [
          {
            scope,
            jsValue,
            message: \`value is neither of the \${numberOfTypes} union types\`,
          },
        ];
  }

  function flatten<T>(arr: T[][]): T[] {
    return arr.reduce((r, a) => [...r, ...a], []);
  }

  const refs: { [refId: string]: (jsValue: unknown) => ValidationError[] } = {};
  refs[\`Something\`] = (jsValue: unknown): ValidationError[] =>
    refs[\`Something\`](jsValue);

  return [...validateTypeOf(\\"number\\", \\"a\\", a)];
}

export async function handler(event: any) {
  const body = JSON.parse(
    event.body === null || event.body === undefined ? \\"{}\\" : event.body
  );

  const inputValidationResult = validate(body.a);

  if (inputValidationResult.length) {
    return {
      statusCode: 400,
      body: JSON.stringify(
        new InvalidInputError(inputValidationResult),
        null,
        4
      ),
      headers: {
        \\"Content-Type\\": \\"application/json\\",
      },
    };
  }

  try {
    const result = await testFunction(body.a);

    return {
      statusCode: 200,
      body: result && JSON.stringify(result, null, 4),
      headers: {
        \\"Content-Type\\": \\"application/json\\",
      },
    };
  } catch (e) {
    return {
      statusCode: 500,
      body: JSON.stringify(e, null, 4),
      headers: {
        \\"Content-Type\\": \\"application/json\\",
      },
    };
  }
}

export async function rpc_testFunction(a: number): Promise<string> {
  const inputValidationResult = validate(a);

  if (inputValidationResult.length) {
    throw new InvalidInputError(inputValidationResult);
  }

  const result = await testFunction(a);
  return result;
}
"
`;

exports[`literal returnType 1`] = `
"import { testFunction } from \\"../samen\\";

export interface ValidationError {
  scope: string;
  jsValue: unknown;
  message: string;
}

class InvalidInputError extends Error {
  errorCode = \\"INVALID_INPUT_ERROR\\";
  constructor(private readonly errors: ValidationError[]) {
    super(\\"Invalid input for RPC\\");
  }
}

export function validate(): ValidationError[] {
  function validateTypeOf(
    typeString: \\"string\\" | \\"number\\" | \\"boolean\\",
    scope: string,
    jsValue: unknown
  ): ValidationError[] {
    return typeof jsValue === typeString
      ? []
      : [{ scope, jsValue, message: \`value is not a \${typeString}\` }];
  }

  function validateValueOf(
    possibleValues: any[],
    scope: string,
    jsValue: any
  ): ValidationError[] {
    return possibleValues.includes(jsValue)
      ? []
      : [
          {
            scope,
            jsValue,
            message: \`value must be one of: \${possibleValues
              .map((v) => \`\\"\${v}\\"\`)
              .join(\\", \\")}\`,
          },
        ];
  }

  function validateDate(scope: string, jsValue: unknown): ValidationError[] {
    const r = /^d{4}-dd-ddTdd:dd:dd(.d+)?(([+-]dd:dd)|Z)?$/i;
    const isValidDate =
      jsValue instanceof Date ||
      (typeof jsValue === \\"string\\" &&
        r.test(jsValue) &&
        !isNaN(Date.parse(jsValue)));
    return isValidDate
      ? []
      : [{ scope, jsValue, message: \\"value must be a date\\" }];
  }

  function validateNull(scope: string, jsValue: unknown): ValidationError[] {
    return jsValue === null
      ? []
      : [{ scope, jsValue, message: \\"value must be null\\" }];
  }

  function validateUndefined(
    scope: string,
    jsValue: unknown
  ): ValidationError[] {
    return jsValue === undefined
      ? []
      : [{ scope, jsValue, message: \\"value must be undefined\\" }];
  }

  function validateObject(scope: string, jsValue: unknown): ValidationError[] {
    return jsValue !== null && typeof jsValue === \\"object\\"
      ? []
      : [{ scope, jsValue, message: \\"value must be an object\\" }];
  }

  function validateArray(scope: string, jsValue: unknown): ValidationError[] {
    return Array.isArray(jsValue)
      ? []
      : [{ scope, jsValue, message: \\"value must be an array\\" }];
  }

  function validateTuple(
    length: number,
    scope: string,
    jsValue: unknown
  ): ValidationError[] {
    return Array.isArray(jsValue) && jsValue.length === length
      ? []
      : [
          {
            scope,
            jsValue,
            message: \`value must be a tuple of length \${length}\`,
          },
        ];
  }

  function validateOneOfTypes(
    numberOfTypes: number,
    scope: string,
    jsValue: unknown,
    isValid: boolean
  ): ValidationError[] {
    return isValid
      ? []
      : [
          {
            scope,
            jsValue,
            message: \`value is neither of the \${numberOfTypes} union types\`,
          },
        ];
  }

  function flatten<T>(arr: T[][]): T[] {
    return arr.reduce((r, a) => [...r, ...a], []);
  }

  const refs: { [refId: string]: (jsValue: unknown) => ValidationError[] } = {};
  refs[\`Something\`] = (jsValue: unknown): ValidationError[] =>
    refs[\`Something\`](jsValue);

  return [];
}

export async function handler(event: any) {
  const body = JSON.parse(
    event.body === null || event.body === undefined ? \\"{}\\" : event.body
  );

  const inputValidationResult = validate();

  if (inputValidationResult.length) {
    return {
      statusCode: 400,
      body: JSON.stringify(
        new InvalidInputError(inputValidationResult),
        null,
        4
      ),
      headers: {
        \\"Content-Type\\": \\"application/json\\",
      },
    };
  }

  try {
    const result = await testFunction();

    return {
      statusCode: 200,
      body: result && JSON.stringify(result, null, 4),
      headers: {
        \\"Content-Type\\": \\"application/json\\",
      },
    };
  } catch (e) {
    return {
      statusCode: 500,
      body: JSON.stringify(e, null, 4),
      headers: {
        \\"Content-Type\\": \\"application/json\\",
      },
    };
  }
}

export async function rpc_testFunction(): Promise<string> {
  const inputValidationResult = validate();

  if (inputValidationResult.length) {
    throw new InvalidInputError(inputValidationResult);
  }

  const result = await testFunction();
  return result;
}
"
`;

exports[`ref parameter 1`] = `
"import { testFunction } from \\"../samen\\";

export interface ValidationError {
  scope: string;
  jsValue: unknown;
  message: string;
}

class InvalidInputError extends Error {
  errorCode = \\"INVALID_INPUT_ERROR\\";
  constructor(private readonly errors: ValidationError[]) {
    super(\\"Invalid input for RPC\\");
  }
}

export function validate(a: unknown): ValidationError[] {
  function validateTypeOf(
    typeString: \\"string\\" | \\"number\\" | \\"boolean\\",
    scope: string,
    jsValue: unknown
  ): ValidationError[] {
    return typeof jsValue === typeString
      ? []
      : [{ scope, jsValue, message: \`value is not a \${typeString}\` }];
  }

  function validateValueOf(
    possibleValues: any[],
    scope: string,
    jsValue: any
  ): ValidationError[] {
    return possibleValues.includes(jsValue)
      ? []
      : [
          {
            scope,
            jsValue,
            message: \`value must be one of: \${possibleValues
              .map((v) => \`\\"\${v}\\"\`)
              .join(\\", \\")}\`,
          },
        ];
  }

  function validateDate(scope: string, jsValue: unknown): ValidationError[] {
    const r = /^d{4}-dd-ddTdd:dd:dd(.d+)?(([+-]dd:dd)|Z)?$/i;
    const isValidDate =
      jsValue instanceof Date ||
      (typeof jsValue === \\"string\\" &&
        r.test(jsValue) &&
        !isNaN(Date.parse(jsValue)));
    return isValidDate
      ? []
      : [{ scope, jsValue, message: \\"value must be a date\\" }];
  }

  function validateNull(scope: string, jsValue: unknown): ValidationError[] {
    return jsValue === null
      ? []
      : [{ scope, jsValue, message: \\"value must be null\\" }];
  }

  function validateUndefined(
    scope: string,
    jsValue: unknown
  ): ValidationError[] {
    return jsValue === undefined
      ? []
      : [{ scope, jsValue, message: \\"value must be undefined\\" }];
  }

  function validateObject(scope: string, jsValue: unknown): ValidationError[] {
    return jsValue !== null && typeof jsValue === \\"object\\"
      ? []
      : [{ scope, jsValue, message: \\"value must be an object\\" }];
  }

  function validateArray(scope: string, jsValue: unknown): ValidationError[] {
    return Array.isArray(jsValue)
      ? []
      : [{ scope, jsValue, message: \\"value must be an array\\" }];
  }

  function validateTuple(
    length: number,
    scope: string,
    jsValue: unknown
  ): ValidationError[] {
    return Array.isArray(jsValue) && jsValue.length === length
      ? []
      : [
          {
            scope,
            jsValue,
            message: \`value must be a tuple of length \${length}\`,
          },
        ];
  }

  function validateOneOfTypes(
    numberOfTypes: number,
    scope: string,
    jsValue: unknown,
    isValid: boolean
  ): ValidationError[] {
    return isValid
      ? []
      : [
          {
            scope,
            jsValue,
            message: \`value is neither of the \${numberOfTypes} union types\`,
          },
        ];
  }

  function flatten<T>(arr: T[][]): T[] {
    return arr.reduce((r, a) => [...r, ...a], []);
  }

  const refs: { [refId: string]: (jsValue: unknown) => ValidationError[] } = {};
  refs[\`Something\`] = (jsValue: unknown): ValidationError[] =>
    refs[\`Something\`](jsValue);

  return [...refs[\`Something\`](a)];
}

export async function handler(event: any) {
  const body = JSON.parse(
    event.body === null || event.body === undefined ? \\"{}\\" : event.body
  );

  const inputValidationResult = validate(body.a);

  if (inputValidationResult.length) {
    return {
      statusCode: 400,
      body: JSON.stringify(
        new InvalidInputError(inputValidationResult),
        null,
        4
      ),
      headers: {
        \\"Content-Type\\": \\"application/json\\",
      },
    };
  }

  try {
    const result = await testFunction(body.a);

    return {
      statusCode: 200,
      body: null,
      headers: {
        \\"Content-Type\\": \\"application/json\\",
      },
    };
  } catch (e) {
    return {
      statusCode: 500,
      body: JSON.stringify(e, null, 4),
      headers: {
        \\"Content-Type\\": \\"application/json\\",
      },
    };
  }
}

export async function rpc_testFunction(a: Something): Promise<void> {
  const inputValidationResult = validate(a);

  if (inputValidationResult.length) {
    throw new InvalidInputError(inputValidationResult);
  }

  const result = await testFunction(a);
  return result;
}
"
`;

exports[`ref parameter and return type 1`] = `
"import { testFunction } from \\"../samen\\";

export interface ValidationError {
  scope: string;
  jsValue: unknown;
  message: string;
}

class InvalidInputError extends Error {
  errorCode = \\"INVALID_INPUT_ERROR\\";
  constructor(private readonly errors: ValidationError[]) {
    super(\\"Invalid input for RPC\\");
  }
}

export function validate(a: unknown): ValidationError[] {
  function validateTypeOf(
    typeString: \\"string\\" | \\"number\\" | \\"boolean\\",
    scope: string,
    jsValue: unknown
  ): ValidationError[] {
    return typeof jsValue === typeString
      ? []
      : [{ scope, jsValue, message: \`value is not a \${typeString}\` }];
  }

  function validateValueOf(
    possibleValues: any[],
    scope: string,
    jsValue: any
  ): ValidationError[] {
    return possibleValues.includes(jsValue)
      ? []
      : [
          {
            scope,
            jsValue,
            message: \`value must be one of: \${possibleValues
              .map((v) => \`\\"\${v}\\"\`)
              .join(\\", \\")}\`,
          },
        ];
  }

  function validateDate(scope: string, jsValue: unknown): ValidationError[] {
    const r = /^d{4}-dd-ddTdd:dd:dd(.d+)?(([+-]dd:dd)|Z)?$/i;
    const isValidDate =
      jsValue instanceof Date ||
      (typeof jsValue === \\"string\\" &&
        r.test(jsValue) &&
        !isNaN(Date.parse(jsValue)));
    return isValidDate
      ? []
      : [{ scope, jsValue, message: \\"value must be a date\\" }];
  }

  function validateNull(scope: string, jsValue: unknown): ValidationError[] {
    return jsValue === null
      ? []
      : [{ scope, jsValue, message: \\"value must be null\\" }];
  }

  function validateUndefined(
    scope: string,
    jsValue: unknown
  ): ValidationError[] {
    return jsValue === undefined
      ? []
      : [{ scope, jsValue, message: \\"value must be undefined\\" }];
  }

  function validateObject(scope: string, jsValue: unknown): ValidationError[] {
    return jsValue !== null && typeof jsValue === \\"object\\"
      ? []
      : [{ scope, jsValue, message: \\"value must be an object\\" }];
  }

  function validateArray(scope: string, jsValue: unknown): ValidationError[] {
    return Array.isArray(jsValue)
      ? []
      : [{ scope, jsValue, message: \\"value must be an array\\" }];
  }

  function validateTuple(
    length: number,
    scope: string,
    jsValue: unknown
  ): ValidationError[] {
    return Array.isArray(jsValue) && jsValue.length === length
      ? []
      : [
          {
            scope,
            jsValue,
            message: \`value must be a tuple of length \${length}\`,
          },
        ];
  }

  function validateOneOfTypes(
    numberOfTypes: number,
    scope: string,
    jsValue: unknown,
    isValid: boolean
  ): ValidationError[] {
    return isValid
      ? []
      : [
          {
            scope,
            jsValue,
            message: \`value is neither of the \${numberOfTypes} union types\`,
          },
        ];
  }

  function flatten<T>(arr: T[][]): T[] {
    return arr.reduce((r, a) => [...r, ...a], []);
  }

  const refs: { [refId: string]: (jsValue: unknown) => ValidationError[] } = {};
  refs[\`Something\`] = (jsValue: unknown): ValidationError[] =>
    refs[\`Something\`](jsValue);

  return [...refs[\`Something\`](a)];
}

export async function handler(event: any) {
  const body = JSON.parse(
    event.body === null || event.body === undefined ? \\"{}\\" : event.body
  );

  const inputValidationResult = validate(body.a);

  if (inputValidationResult.length) {
    return {
      statusCode: 400,
      body: JSON.stringify(
        new InvalidInputError(inputValidationResult),
        null,
        4
      ),
      headers: {
        \\"Content-Type\\": \\"application/json\\",
      },
    };
  }

  try {
    const result = await testFunction(body.a);

    return {
      statusCode: 200,
      body: result && JSON.stringify(result, null, 4),
      headers: {
        \\"Content-Type\\": \\"application/json\\",
      },
    };
  } catch (e) {
    return {
      statusCode: 500,
      body: JSON.stringify(e, null, 4),
      headers: {
        \\"Content-Type\\": \\"application/json\\",
      },
    };
  }
}

export async function rpc_testFunction(a: Something): Promise<Something> {
  const inputValidationResult = validate(a);

  if (inputValidationResult.length) {
    throw new InvalidInputError(inputValidationResult);
  }

  const result = await testFunction(a);
  return result;
}
"
`;

exports[`ref returnType 1`] = `
"import { testFunction } from \\"../samen\\";

export interface ValidationError {
  scope: string;
  jsValue: unknown;
  message: string;
}

class InvalidInputError extends Error {
  errorCode = \\"INVALID_INPUT_ERROR\\";
  constructor(private readonly errors: ValidationError[]) {
    super(\\"Invalid input for RPC\\");
  }
}

export function validate(): ValidationError[] {
  function validateTypeOf(
    typeString: \\"string\\" | \\"number\\" | \\"boolean\\",
    scope: string,
    jsValue: unknown
  ): ValidationError[] {
    return typeof jsValue === typeString
      ? []
      : [{ scope, jsValue, message: \`value is not a \${typeString}\` }];
  }

  function validateValueOf(
    possibleValues: any[],
    scope: string,
    jsValue: any
  ): ValidationError[] {
    return possibleValues.includes(jsValue)
      ? []
      : [
          {
            scope,
            jsValue,
            message: \`value must be one of: \${possibleValues
              .map((v) => \`\\"\${v}\\"\`)
              .join(\\", \\")}\`,
          },
        ];
  }

  function validateDate(scope: string, jsValue: unknown): ValidationError[] {
    const r = /^d{4}-dd-ddTdd:dd:dd(.d+)?(([+-]dd:dd)|Z)?$/i;
    const isValidDate =
      jsValue instanceof Date ||
      (typeof jsValue === \\"string\\" &&
        r.test(jsValue) &&
        !isNaN(Date.parse(jsValue)));
    return isValidDate
      ? []
      : [{ scope, jsValue, message: \\"value must be a date\\" }];
  }

  function validateNull(scope: string, jsValue: unknown): ValidationError[] {
    return jsValue === null
      ? []
      : [{ scope, jsValue, message: \\"value must be null\\" }];
  }

  function validateUndefined(
    scope: string,
    jsValue: unknown
  ): ValidationError[] {
    return jsValue === undefined
      ? []
      : [{ scope, jsValue, message: \\"value must be undefined\\" }];
  }

  function validateObject(scope: string, jsValue: unknown): ValidationError[] {
    return jsValue !== null && typeof jsValue === \\"object\\"
      ? []
      : [{ scope, jsValue, message: \\"value must be an object\\" }];
  }

  function validateArray(scope: string, jsValue: unknown): ValidationError[] {
    return Array.isArray(jsValue)
      ? []
      : [{ scope, jsValue, message: \\"value must be an array\\" }];
  }

  function validateTuple(
    length: number,
    scope: string,
    jsValue: unknown
  ): ValidationError[] {
    return Array.isArray(jsValue) && jsValue.length === length
      ? []
      : [
          {
            scope,
            jsValue,
            message: \`value must be a tuple of length \${length}\`,
          },
        ];
  }

  function validateOneOfTypes(
    numberOfTypes: number,
    scope: string,
    jsValue: unknown,
    isValid: boolean
  ): ValidationError[] {
    return isValid
      ? []
      : [
          {
            scope,
            jsValue,
            message: \`value is neither of the \${numberOfTypes} union types\`,
          },
        ];
  }

  function flatten<T>(arr: T[][]): T[] {
    return arr.reduce((r, a) => [...r, ...a], []);
  }

  const refs: { [refId: string]: (jsValue: unknown) => ValidationError[] } = {};
  refs[\`Something\`] = (jsValue: unknown): ValidationError[] =>
    refs[\`Something\`](jsValue);

  return [];
}

export async function handler(event: any) {
  const body = JSON.parse(
    event.body === null || event.body === undefined ? \\"{}\\" : event.body
  );

  const inputValidationResult = validate();

  if (inputValidationResult.length) {
    return {
      statusCode: 400,
      body: JSON.stringify(
        new InvalidInputError(inputValidationResult),
        null,
        4
      ),
      headers: {
        \\"Content-Type\\": \\"application/json\\",
      },
    };
  }

  try {
    const result = await testFunction();

    return {
      statusCode: 200,
      body: result && JSON.stringify(result, null, 4),
      headers: {
        \\"Content-Type\\": \\"application/json\\",
      },
    };
  } catch (e) {
    return {
      statusCode: 500,
      body: JSON.stringify(e, null, 4),
      headers: {
        \\"Content-Type\\": \\"application/json\\",
      },
    };
  }
}

export async function rpc_testFunction(): Promise<Something> {
  const inputValidationResult = validate();

  if (inputValidationResult.length) {
    throw new InvalidInputError(inputValidationResult);
  }

  const result = await testFunction();
  return result;
}
"
`;

exports[`ref-array returnType 1`] = `
"import { testFunction } from \\"../samen\\";

export interface ValidationError {
  scope: string;
  jsValue: unknown;
  message: string;
}

class InvalidInputError extends Error {
  errorCode = \\"INVALID_INPUT_ERROR\\";
  constructor(private readonly errors: ValidationError[]) {
    super(\\"Invalid input for RPC\\");
  }
}

export function validate(): ValidationError[] {
  function validateTypeOf(
    typeString: \\"string\\" | \\"number\\" | \\"boolean\\",
    scope: string,
    jsValue: unknown
  ): ValidationError[] {
    return typeof jsValue === typeString
      ? []
      : [{ scope, jsValue, message: \`value is not a \${typeString}\` }];
  }

  function validateValueOf(
    possibleValues: any[],
    scope: string,
    jsValue: any
  ): ValidationError[] {
    return possibleValues.includes(jsValue)
      ? []
      : [
          {
            scope,
            jsValue,
            message: \`value must be one of: \${possibleValues
              .map((v) => \`\\"\${v}\\"\`)
              .join(\\", \\")}\`,
          },
        ];
  }

  function validateDate(scope: string, jsValue: unknown): ValidationError[] {
    const r = /^d{4}-dd-ddTdd:dd:dd(.d+)?(([+-]dd:dd)|Z)?$/i;
    const isValidDate =
      jsValue instanceof Date ||
      (typeof jsValue === \\"string\\" &&
        r.test(jsValue) &&
        !isNaN(Date.parse(jsValue)));
    return isValidDate
      ? []
      : [{ scope, jsValue, message: \\"value must be a date\\" }];
  }

  function validateNull(scope: string, jsValue: unknown): ValidationError[] {
    return jsValue === null
      ? []
      : [{ scope, jsValue, message: \\"value must be null\\" }];
  }

  function validateUndefined(
    scope: string,
    jsValue: unknown
  ): ValidationError[] {
    return jsValue === undefined
      ? []
      : [{ scope, jsValue, message: \\"value must be undefined\\" }];
  }

  function validateObject(scope: string, jsValue: unknown): ValidationError[] {
    return jsValue !== null && typeof jsValue === \\"object\\"
      ? []
      : [{ scope, jsValue, message: \\"value must be an object\\" }];
  }

  function validateArray(scope: string, jsValue: unknown): ValidationError[] {
    return Array.isArray(jsValue)
      ? []
      : [{ scope, jsValue, message: \\"value must be an array\\" }];
  }

  function validateTuple(
    length: number,
    scope: string,
    jsValue: unknown
  ): ValidationError[] {
    return Array.isArray(jsValue) && jsValue.length === length
      ? []
      : [
          {
            scope,
            jsValue,
            message: \`value must be a tuple of length \${length}\`,
          },
        ];
  }

  function validateOneOfTypes(
    numberOfTypes: number,
    scope: string,
    jsValue: unknown,
    isValid: boolean
  ): ValidationError[] {
    return isValid
      ? []
      : [
          {
            scope,
            jsValue,
            message: \`value is neither of the \${numberOfTypes} union types\`,
          },
        ];
  }

  function flatten<T>(arr: T[][]): T[] {
    return arr.reduce((r, a) => [...r, ...a], []);
  }

  const refs: { [refId: string]: (jsValue: unknown) => ValidationError[] } = {};
  refs[\`Something\`] = (jsValue: unknown): ValidationError[] =>
    refs[\`Something\`](jsValue);

  return [];
}

export async function handler(event: any) {
  const body = JSON.parse(
    event.body === null || event.body === undefined ? \\"{}\\" : event.body
  );

  const inputValidationResult = validate();

  if (inputValidationResult.length) {
    return {
      statusCode: 400,
      body: JSON.stringify(
        new InvalidInputError(inputValidationResult),
        null,
        4
      ),
      headers: {
        \\"Content-Type\\": \\"application/json\\",
      },
    };
  }

  try {
    const result = await testFunction();

    return {
      statusCode: 200,
      body: result && JSON.stringify(result, null, 4),
      headers: {
        \\"Content-Type\\": \\"application/json\\",
      },
    };
  } catch (e) {
    return {
      statusCode: 500,
      body: JSON.stringify(e, null, 4),
      headers: {
        \\"Content-Type\\": \\"application/json\\",
      },
    };
  }
}

export async function rpc_testFunction(): Promise<Something[]> {
  const inputValidationResult = validate();

  if (inputValidationResult.length) {
    throw new InvalidInputError(inputValidationResult);
  }

  const result = await testFunction();
  return result;
}
"
`;
